node/src/components/linear_chain/state.rs

```diff
@@ -316,137 +465,30 @@ impl LinearChain {
-                ..
-            } = *new_fs;
-            warn!(
-                validator = %public_key,
-                %block_hash,
-                "Received a signature from a validator that is not bonded."
-            );
-            // TODO: Disconnect from the sender.
+        if !self.is_new(&fs) {
+            debug!(block_hash=%fs.block_hash, public_key=%fs.public_key,
+                "finality signature is already known");
             return vec![];
         }
-
-        self.pending_finality_signatures
-            .mark_bonded(new_fs.public_key.clone(), new_fs.block_hash);
-
-        match maybe_known_signatures
-            .or_else(|| self.get_signatures(&new_fs.block_hash).map(Box::new))
-        {
-            None => {
-                // Unknown block but validator is bonded.
-                // We should finalize the same block eventually. Either in this or in the
-                // next eras. New signature is already cached for later.
-                vec![]
-            }
-            Some(mut known_signatures) => {
-                // New finality signature from a bonded validator.
-                known_signatures.insert_proof(new_fs.public_key.clone(), new_fs.signature);
-                // Cache the results in case we receive the same finality signature before we
-                // manage to store it in the database.
-                self.cache_signatures(*known_signatures.clone());
-                debug!(hash = %known_signatures.block_hash, "storing finality signatures");
-                // Announce new finality signatures for other components to pick up.
-                let mut outcomes = vec![Outcome::AnnounceSignature(new_fs.clone())];
-                if let Some(signature) = self.remove_from_pending_fs(&*new_fs) {
-                    // This shouldn't return `None` as we added the `fs` to the pending collection
-                    // when we received it. If it _is_ `None` then a concurrent
-                    // flow must have already removed it. If it's a signature
-                    // created by this node, gossip it.
-                    if signature.is_local() {
-                        outcomes.push(Outcome::Gossip(new_fs.clone()));
-                    }
-                };
-                let should_upgrade = self.should_upgrade(&*known_signatures);
-                outcomes.push(Outcome::StoreBlockSignatures(
-                    *known_signatures,
-                    should_upgrade,
-                ));
-                outcomes
-            }
+        // Not found in the cache, look in the storage.
+        // Add as pending check so that if we receive the same signature again we don't repeat the
+        // steps.
+        if !self.add_pending_finality_signature(*fs.clone(), gossiped) {
+            // If we did not add the signature it means it's either incorrect or we already
+            // know it.
+            return vec![];
         }
+        vec![Outcome::LoadSignatures(fs)]
```

Comment:
So we don't check for cached signatures here anymore?

Reply:
We do in `self.is_new` that comes earlier, right after `self.is_pending`.